### PR TITLE
8305093: Linker cache should not take layout names into account

### PR DESCRIPTION
**Pull request by JornVernee**
Strip names from the function descriptors used in linkage requests. The names are not important for classification, and this will allow the cached MethodHandle/UpcallStubFactory to be shared across linkage request with the same structure, but with layouts that have different names.

I also have another implementation that avoids copying all the layouts, and instead re-implemented equals and hashCode in the LinkRequest wrapper: https://github.com/openjdk/jdk/compare/pr/13079...JornVernee:jdk:NoNameCache But, this seems too complex. Linking code is not very performance sensitive, so I instead went with a simpler approach in this patch.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8305093](https://bugs.openjdk.org/browse/JDK-8305093): Linker cache should not take layout names into account

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**) ⚠️ Review applies to [cc7c4ebf](https://git.openjdk.org/jdk/pull/13562/files/cc7c4ebfd4f0fb88278e654a2bff16f928da95dc)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13562/head:pull/13562` \
`$ git checkout pull/13562`

Update a local copy of the PR: \
`$ git checkout pull/13562` \
`$ git pull https://git.openjdk.org/jdk.git pull/13562/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13562`

View PR using the GUI difftool: \
`$ git pr show -t 13562`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13562.diff">https://git.openjdk.org/jdk/pull/13562.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13562#issuecomment-1517102928)

### CODE REVIEW

File: src/java.base/share/classes/jdk/internal/foreign/abi/AbstractLinker.java [hunk 1]
```diff
@@ -108,7 +117,6 @@ private static void checkHasNaturalAlignment(FunctionDescriptor descriptor) {
         descriptor.returnLayout().ifPresent(AbstractLinker::checkHasNaturalAlignmentRecursive);
         descriptor.argumentLayouts().forEach(AbstractLinker::checkHasNaturalAlignmentRecursive);
     }
-
     private static void checkHasNaturalAlignmentRecursive(MemoryLayout layout) {
```
Comment (ExE-Boss) [inline src/java.base/share/classes/jdk/internal/foreign/abi/AbstractLinker.java:120]: **Nit:** restore deleted line between methods:
```suggestion
    }

    private static void checkHasNaturalAlignmentRecursive(MemoryLayout layout) {
```

File: src/java.base/share/classes/jdk/internal/foreign/abi/AbstractLinker.java [hunk 2]
```diff
@@ -125,4 +134,30 @@ private static void checkHasNaturalAlignment(MemoryLayout layout) {
             throw new IllegalArgumentException("Layout bit alignment must be natural alignment: " + layout);
         }
     }
+
+    private static MemoryLayout stripNames(MemoryLayout ml) {
+        // we don't care about transferring alignment and byte order here
+        // since the linker already restricts those such that they will always be the same
+        return switch (ml) {
+            case StructLayout sl -> MemoryLayout.structLayout(stripNames(sl.memberLayouts()));
```
Comment (mcimadamore) [inline src/java.base/share/classes/jdk/internal/foreign/abi/AbstractLinker.java:142]: shouldn't names be also stripped from enclosing struct/union/sequence?

Comment (JornVernee) [reply src/java.base/share/classes/jdk/internal/foreign/abi/AbstractLinker.java:142]: That should already happen because we're recreating the layout here.

**Review by mcimadamore — APPROVED** [review]

File: test/jdk/java/foreign/TestByteBuffer.java [hunk 3]
```diff
@@ -290,26 +285,65 @@ public void testMappedSegmentOffset() throws Throwable {
...
+                    MemorySegment segment = fileChannel.map(mapOption, 0L, 32L, arena);
+                    segment.force();
+                    segment.load();
+                    segment.isLoaded();
+                    segment.unload();
+                    ByteBuffer byteBuffer = segment.asByteBuffer();
+                    assertEquals(byteBuffer.capacity(), segment.byteSize());
+                    assertEquals(byteBuffer.isReadOnly(), segment.isReadOnly());
+                    assertTrue(byteBuffer.isDirect());
+                } catch(IOException e) {
```
Comment (turbanoff) [inline test/jdk/java/foreign/TestByteBuffer.java:335]: nit
```suggestion
                } catch (IOException e) {
```

File: src/java.base/share/classes/jdk/internal/foreign/abi/AbstractLinker.java [hunk 4]
```diff
@@ -121,4 +134,30 @@ private static void checkHasNaturalAlignment(MemoryLayout layout) {
...
+            case UnionLayout ul -> MemoryLayout.unionLayout(stripNames(ul.memberLayouts()));
+            case SequenceLayout sl -> MemoryLayout.sequenceLayout(sl.elementCount(), stripNames(sl.elementLayout()));
+            case AddressLayout al -> al.targetLayout()
+                    .map(tl -> al.withoutName().withTargetLayout(stripNames(tl)))
+                    .orElse(al.withoutName());
```
Comment (ExE-Boss) [inline src/java.base/share/classes/jdk/internal/foreign/abi/AbstractLinker.java:147]: This can use [`Optional::orElseGet`]:
```suggestion
                    .orElseGet(al::withoutName);
```

[`Optional::orElseGet`]: https://docs.oracle.com/en/java/javase/20/docs/api/java.base/java/util/Optional.html#orElseGet%28java.util.function.Supplier%29

**Pull request closed without merging**